pygmt/base_plotting.py

```diff
@@ -1211,6 +1211,129 @@ def basemap(self, **kwargs):
+            rectangle in projected coordinates and optionally
+            append **+u**\ *unit* [Default coordinate unit is meter (e)].
+
+            Append **+w**\ *width*\ [/*height*] of bounding rectangle or box
+            in plot coordinates (inches, cm, etc.). By default, the anchor
+            point on the scale is assumed to be the bottom left corner (BL),
+            but this can be changed by appending **+j** followed by a 2-char
+            justification code *justify*.
+            **Note**: If **j** is used then *justify* defaults to the same
+            as *refpoint*, if **J** is used then *justify* defaults to the
+            mirror opposite of *refpoint*. Specify inset box attributes via
+            the ``border`` option [outline only].
+        border : str or bool
+            [**+c**\ *clearances*][**+g**\ *fill*][**+i**\ [[*gap*/]\
+            *pen*]][**+p**\ [*pen*]][**+r**\ [*radius*]][**+s**\
+            [[*dx*/*dy*/][*shade*]]]
+
+            If passed ``True``, this draws a rectangular border around the map
+            inset using the default pen; specify a different pen
+            with **+p**\ *pen*. Add **+g**\ *fill* to fill the logo box
+            [Default is no fill].
+            Append **+c**\ *clearance*  where *clearance* is either
+            *gap*, *xgap*\ /\ *ygap*, or *lgap*\ /\ *rgap*\ /\ *bgap*\ /\
+            *tgap* where these items are uniform, separate in x- and
+            y-direction, or individual side spacings between logo and border.
+            Append **+i** to draw a secondary, inner border as well. We use a
+            uniform *gap* between borders of 2\ **p** and the default pen
+            unless other values are specified. Append **+r** to draw rounded
+            rectangular borders instead, with a 6\ **p** corner radius. You
+            can override this radius by appending another value. Append
+            **+s** to draw an offset background shaded region. Here, *dx*/*dy*
+            indicates the shift relative to the foreground frame
+            [4\ **p**/-4\ **p**] and *shade* sets the fill style to use for
+            shading [Default is gray50].
+        margin : int or str or list
+            This is clearance that is added around the inside of the inset.
+            Plotting will take place within the inner region only. The margins
+            can be a single value, a pair of values separated (for setting
+            separate horizontal and vertical margins), or the full set of four
+            margins (for setting separate left, right, bottom, and top
+            margins). When passing multiple values, it can be either a list or
+            a string with the values separated by forward
+            slashes [Default is no margins].
+        {V}
+        """
+        kwargs = self._preprocess(**kwargs)
+        with Session() as lib:
+            lib.call_module("inset", "begin " + build_arg_string(kwargs))
```

Comment:
You'll need to update/consolidate the documentation and change the tests to use a `with:` block, but this should get you started:

```suggestion
        with Session() as lib:
            try:
                lib.call_module("inset", f"begin {build_arg_string(kwargs)}")
                yield
            finally:
                lib.call_module("inset", "end")
```

and this should work:

```python
def test_inset():
    fig = Figure()
    fig.coast(
        region="MG+r2", land="brown", water="lightblue", shorelines="thin", frame="a"
    )
    with fig.inset(location="jTL+w3.5c+o0.2c", margin=0, border="+pgreen"):
        fig.coast(
            region="g", projection="G47/-20/4c", land="gray", water="white", E="MG+gred"
        )
    return fig
```
![inset fig](https://user-images.githubusercontent.com/23487320/105404553-fcb89a00-5c8e-11eb-92f1-ee8b8310a461.png)

Reply:
@weiji14 Thanks; I'll work on this tomorrow morning!

Also, not that I'm opposed to creating branches off of the pygmt repo, but what's the advantage of that over just branching off my personal fork (which I can quickly update with `git pull upstream master`)?

Reply:
> @weiji14 Thanks; I'll work on this tomorrow morning!

Great, and try and have a think about adding a gallery example in this PR if possible.

> Also, not that I'm opposed to creating branches off of the pygmt repo, but what's the advantage of that over just branching off my personal fork (which I can quickly update with `git pull upstream master`)?

1. Easier collaboration. Other `pygmt` maintainers can push to branches on GenericMappingTools/pygmt branch more easily than to your fork.
2. The Vercel link for the PR will be shorter and more memorable. Instead of https://pygmt-git-fork-willschlitzer-wrap-inset.gmt.vercel.app, it would be something like https://pygmt-git-wrap-inset.gmt.vercel.app
3. Prevent an `unknown-repository` from showing up if you delete your branch https://softwareengineering.stackexchange.com/questions/61577/should-i-keep-my-github-forked-repositories-around-forever

Personally I use a Git GUI called [Gitkraken](https://www.gitkraken.com) to manage all the repos I work on. For the `pygmt` project, I set GenericMappingTools/pygmt as `upstream`, and my personal fork (which I hardly use anymore) as `origin`. But yeah, it's up to you to find a workflow you're comfortable with, everyone does things slightly differently.